applications/asset_tracker/src/main.c

```diff
@@ -568,247 +556,91 @@ static void leds_update(struct k_work *work)
 }
 
 /**@brief Send sensor data to nRF Cloud. **/
-static void sensor_data_send(struct nrf_cloud_sensor_data *data)
+static void sensor_data_send(struct cloud_sensor_data *data)
 {
-	int err;
+	int err = 0;
+	struct cloud_data output;
 
-	if (pattern_recording || !atomic_get(&send_data_enable)) {
+	if (!atomic_get(&send_data_enable)) {
 		return;
 	}
 
-	if (data->type == NRF_CLOUD_SENSOR_GPS) {
-		err = nrf_cloud_sensor_data_send(data);
-	} else {
-		err = nrf_cloud_sensor_data_stream(data);
-	}
-
-	if (err) {
-		printk("sensor_data_send failed: %d\n", err);
-		nrf_cloud_error_handler(err);
-	}
-}
-
-/**@brief Callback for user association event received from nRF Cloud. */
-static void on_user_association_req(const struct nrf_cloud_evt *p_evt)
-{
-	if (!pattern_recording) {
-		k_sem_init(&user_assoc_sem, 0, 1);
-		display_state = LEDS_PATTERN_WAIT;
-		pattern_recording = true;
-		buttons_captured = 0;
-		buttons_to_capture = p_evt->param.ua_req.sequence.len;
-
-		printk("Please enter the user association pattern ");
-
-		if (IS_ENABLED(CONFIG_CLOUD_UA_BUTTONS)) {
-			printk("using the buttons and switches\n");
-		} else if (IS_ENABLED(CONFIG_CLOUD_UA_CONSOLE)) {
-			printk("using the console\n");
-			console_init();
-		}
-	}
-}
+	err = cloud_encode_sensor_data(data, &output);
 
-/**@brief Callback for data received event from nRF Cloud. */
-static void on_data_received(const struct nrf_cloud_evt *p_evt)
-{
-	if (memcmp(p_evt->param.data.ptr, CLOUD_LED_ON_STR,
-		strlen(CLOUD_LED_ON_STR)) == 0) {
-		display_state |= CLOUD_LED_MSK;
-	} else if (memcmp(p_evt->param.data.ptr, CLOUD_LED_OFF_STR,
-		strlen(CLOUD_LED_OFF_STR)) == 0) {
-		display_state &= ~CLOUD_LED_MSK;
-	} else {
-		printk("Data not recognised\n");
-	}
-}
+	struct cloud_msg msg = {
+		.buf = output.buf,
+		.len = output.len,
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_TOPIC_MSG
+	};
 
-/**@brief Attach available sensors to nRF Cloud. */
-void sensors_attach(void)
-{
-	int err;
-	struct nrf_cloud_sa_param param;
+	err = cloud_send(cloud_backend, &msg);
 
-	for (u8_t i = 0; i < ARRAY_SIZE(available_sensors); i++) {
-		param.sensor_type = available_sensors[i];
+	k_free(output.buf);
```

Comment:
I'd consider adding an API to `cloud_codec` for releasing allocated data (something like `cloud_release_data(struct cloud_data *data)`). The app should not need to know what exactly it has to release.

Reply:
Good idea, added the suggested API